File: iTSoft.CRM.Domain/Services/Process/DashboardService.cs
using Dapper;
using iTSoft.CRM.Data.Core;
using iTSoft.CRM.Domain.Models.ViewModel;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace iTSoft.CRM.Domain.Services.Process
{

    public interface IDashboardService
    {
         Task<List<LeadSourceDashboardViewModel>> GetLeadSourceDashboard(DashboardSearchParameters searchParameters);

        Task<RevenueTargetDashboardViewModel> GetRevenueTargetDashboard(DashboardSearchParameters searchParameters);

        Task<List<TopNEmployeeDashboardViewModel>> GetTopNEmployeeDashboard(DashboardSearchParameters searchParameters);

        Task<List<DepartmentWiseRevenueDashboardViewModel>> GetDepartmentWiseRevenueDashboard(DashboardSearchParameters searchParameters);

        Task<List<LeadStatusDashboardViewModel>> GetLeadStatusDashboard(DashboardSearchParameters searchParameters);


    }
    public class DashboardService : BaseRepository, IDashboardService
    {
        public const string PROC_LeadSourceDashboard = "PROC_LeadSourceDashboard";
        public const string PROC_RevenueTargetDashboard = "PROC_RevenueTargetDashboard";

        public const string PROC_TopNEmployeeDashboard = "PROC_TopNEmployeeDashboard";
        public const string PROC_DepartmentWiseRevenueDashboard = "PROC_DepartmentWiseRevenueDashboard";
        public const string PROC_StatusWiseLeadDashboard = "PROC_StatusWiseLeadDashboard";

        public async Task<List<DepartmentWiseRevenueDashboardViewModel>> GetDepartmentWiseRevenueDashboard(DashboardSearchParameters searchParameters)
        {
            using (IDbConnection dbConnection = base.GetConnection())
            {
                DynamicParameters param = new DynamicParameters();
                param.Add(nameof(searchParameters.FromDate), searchParameters.FromDate);
                param.Add(nameof(searchParameters.ToDate), searchParameters.ToDate);
                var result = await dbConnection.QueryAsync<DepartmentWiseRevenueDashboardViewModel>(PROC_DepartmentWiseRevenueDashboard, param, commandType: CommandType.StoredProcedure);

                return await Task.FromResult(result.AsList<DepartmentWiseRevenueDashboardViewModel>());
            }
        }

        public async Task<List<LeadSourceDashboardViewModel>> GetLeadSourceDashboard(DashboardSearchParameters searchParameters)
        {
            using (IDbConnection dbConnection = base.GetConnection())
            {
                DynamicParameters param = new DynamicParameters();
                param.Add(nameof(searchParameters.AdvisorId), searchParameters.AdvisorId);
                param.Add(nameof(searchParameters.FromDate), searchParameters.FromDate);
                param.Add(nameof(searchParameters.ToDate), searchParameters.ToDate);
                var result = await dbConnection.QueryAsync<LeadSourceDashboardViewModel>(PROC_LeadSourceDashboard, param, commandType: CommandType.StoredProcedure);

                return await Task.FromResult(result.AsList<LeadSourceDashboardViewModel>());
            }
        }

        public async Task<List<LeadStatusDashboardViewModel>> GetLeadStatusDashboard(DashboardSearchParameters searchParameters)
        {
            using (IDbConnection dbConnection = base.GetConnection())
            {
                DynamicParameters param = new DynamicParameters();
                param.Add(nameof(searchParameters.FromDate), searchParameters.FromDate);
                param.Add(nameof(searchParameters.ToDate), searchParameters.ToDate);
                var result = await dbConnection.QueryAsync<LeadStatusDashboardViewModel>(PROC_StatusWiseLeadDashboard, param, commandType: CommandType.StoredProcedure);
                return await Task.FromResult(result.AsList<LeadStatusDashboardViewModel>());
            }
        }

        public async Task<RevenueTargetDashboardViewModel> GetRevenueTargetDashboard(DashboardSearchParameters searchParameters)
        {
            using (IDbConnection dbConnection = base.GetConnection())
            {
                DynamicParameters param = new DynamicParameters();
                param.Add(nameof(searchParameters.AdvisorId), searchParameters.AdvisorId);
                param.Add(nameof(searchParameters.FromDate), searchParameters.FromDate);
                param.Add(nameof(searchParameters.ToDate), searchParameters.ToDate);
                var result = await dbConnection.QueryAsync<RevenueTargetDashboardViewModel>(PROC_RevenueTargetDashboard, param, commandType: CommandType.StoredProcedure);
                return await Task.FromResult(result.AsList<RevenueTargetDashboardViewModel>().FirstOrDefault());
            }
        }

        public async Task<List<TopNEmployeeDashboardViewModel>> GetTopNEmployeeDashboard(DashboardSearchParameters searchParameters)
        {
            using (IDbConnection dbConnection = base.GetConnection())
            {
                DynamicParameters param = new DynamicParameters();
                param.Add(nameof(searchParameters.FromDate), searchParameters.FromDate);
                param.Add(nameof(searchParameters.ToDate), searchParameters.ToDate);
                param.Add(nameof(searchParameters.NumberOfEmployees), searchParameters.NumberOfEmployees);
                var result = await dbConnection.QueryAsync<TopNEmployeeDashboardViewModel>(PROC_TopNEmployeeDashboard, param, commandType: CommandType.StoredProcedure);
                return await Task.FromResult(result.AsList<TopNEmployeeDashboardViewModel>());
            }
        }
    }
}

File: iTSoft.CRM.UI/src/app/dashboard/dashboard.service.ts
import { Injectable } from "@angular/core";
import { iTCRMSettings } from "../core/models/iTSOFT.iTCRM.Configuration";
import { APIService } from "../_services";

@Injectable()
export class DashboardService {
  
    constructor(private apiService: APIService) { }

    URLGetLeadSourceDashboard: string = iTCRMSettings.Process + "/Dashboard/GetLeadSourceDashboard";
    URLGetLeadStatusDashboard: string = iTCRMSettings.Process + "/Dashboard/GetLeadStatusDashboard";
    URLGetRevenueTargetDashboard: string = iTCRMSettings.Process + "/Dashboard/GetRevenueTargetDashboard";
    URLGetTopNEmployeeDashboard: string = iTCRMSettings.Process + "/Dashboard/GetTopNEmployeeDashboard";
    URLGetDepartmentWiseRevenueDashboard: string = iTCRMSettings.Process + "/Dashboard/GetDepartmentWiseRevenueDashboard";
    
    GetLeadSourceDashboard(searchParam: DashboardSearchParameters) {
        return this.apiService.PostData(this.URLGetLeadSourceDashboard, searchParam);
    }

    GetLeadStatusDashboard(searchParam: DashboardSearchParameters) {
        return this.apiService.PostData(this.URLGetLeadStatusDashboard, searchParam);
    }

    GetRevenueTargetDashboard(searchParam: DashboardSearchParameters) {
        return this.apiService.PostData(this.URLGetRevenueTargetDashboard, searchParam);
    }

    GetTopNEmployeeDashboard(searchParam: DashboardSearchParameters) {
        return this.apiService.PostData(this.URLGetTopNEmployeeDashboard, searchParam);
    }

    GetDepartmentWiseRevenueDashboard(searchParam: DashboardSearchParameters) {
        return this.apiService.PostData(this.URLGetDepartmentWiseRevenueDashboard, searchParam);
    }

}


export class DashboardSearchParameters {
    AdvisorId: number;
    FromDate: Date;
    ToDate: Date;
    NumberOfEmployees: number;
}


export class SeriesDashboardViewModel {
   
    name : string ="";
    series : Array<NameNumber> = [];

}

export class NameValue {
    name: string;
    value:string | number;
}

export class NameNumber {
    name: string;
    value: number;
}

export class TopNEmployeeDashboardViewModel {
    EmployeeName: string;
    TargetAmount: number;
    RevenueGenerated: number;
}

export class DepartmentWiseRevenueDashboardViewModel {
    DepartmentName: string;
    RevenueGenerated: number;
}

export class LeadStatusDashboardViewModel {
    LeadStatusName: string;
    RequestCount: number;
}